### PR TITLE
feat: propagate warnings from composition

### PR DESCRIPTION
**Pull request by JivusAyrus**
## Checklist

- [x] I have discussed my proposed changes in an issue and have received approval to proceed.
- [x] I have followed the coding standards of the project.
- [ ] Tests or benchmarks have been added or updated.
- [ ] Documentation has been updated on [https://github.com/wundergraph/cosmo-docs](https://github.com/wundergraph/cosmo-docs).
- [x] I have read the [Contributors Guide](https://github.com/wundergraph/cosmo/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: cli/src/commands/contract/commands/create.ts [hunk 1]
```diff
@@ -121,6 +124,30 @@ export default (opts: BaseCommandOptions) => {
...
+          pc.bold(pc.white('FEDERATED_GRAPH_NAME')),
+          pc.bold(pc.white('NAMESPACE')),
+          pc.bold(pc.white('FEATURE_FLAG')),
+          pc.bold(pc.white('WARNING_MESSAGE')),
+        ],
+        colWidths: [30, 30, 30, 120],
+        wordWrap: true,
+      });
+
+      console.log(pc.yellow(`We found these composition warnings, while composing the federated graph.`));
```
Comment (Aenimus) [inline cli/src/commands/contract/commands/create.ts:140]: `The following warnings were produced while composing the federated graph:`

File: cli/src/commands/contract/commands/create.ts [hunk 2]
```diff
@@ -28,6 +28,7 @@ export default (opts: BaseCommandOptions) => {
     '--admission-webhook-secret [string]',
     'The admission webhook secret is used to sign requests to the webhook url.',
   );
+  command.option('--suppress-warnings', 'This flag suppresses the composition warnings.');
```
Comment (Aenimus) [inline cli/src/commands/contract/commands/create.ts:31]: `This flag suppresses any warnings produced by composition.`

File: cli/src/commands/contract/commands/update.ts [hunk 3]
```diff
@@ -12,6 +12,7 @@ export default (opts: BaseCommandOptions) => {
   command.argument('<name>', 'The name of the contract graph to update.');
   command.option('-n, --namespace [string]', 'The namespace of the contract update.');
   command.option('--exclude [tags...]', 'Schema elements with these tags will be excluded from the contract schema.');
+  command.option('--suppress-warnings', 'This flag suppresses the composition warnings.');
```
Comment (Aenimus) [inline cli/src/commands/contract/commands/update.ts:15]: Same as above

File: cli/src/commands/contract/commands/update.ts [hunk 4]
```diff
@@ -87,6 +90,30 @@ export default (opts: BaseCommandOptions) => {
...
+          pc.bold(pc.white('FEDERATED_GRAPH_NAME')),
+          pc.bold(pc.white('NAMESPACE')),
+          pc.bold(pc.white('FEATURE_FLAG')),
+          pc.bold(pc.white('WARNING_MESSAGE')),
+        ],
+        colWidths: [30, 30, 30, 120],
+        wordWrap: true,
+      });
+
+      console.log(pc.yellow(`We found these composition warnings, while composing the federated graph.`));
```
Comment (Aenimus) [inline cli/src/commands/contract/commands/update.ts:106]: Same as above

File: cli/src/commands/contract/commands/update.ts [hunk 5]
```diff
@@ -87,6 +90,30 @@ export default (opts: BaseCommandOptions) => {
         process.exit(1);
       }
     }
+
+    if (!options.suppressWarnings && resp.compositionWarnings.length > 0) {
```
Comment (Aenimus) [inline cli/src/commands/contract/commands/update.ts:94]: You could shorten this to `options.suppressWarnings?.length > 0` if you wanted.

File: cli/src/commands/feature-flag/commands/create.ts [hunk 6]
```diff
@@ -29,6 +29,7 @@ export default (opts: BaseCommandOptions) => {
       ' A new feature flag is disabled by default to prevent accidental compositions.',
   );
   command.option('-j, --json', 'Prints to the console in json format instead of table');
+  command.option('--suppress-warnings', 'This flag suppresses the composition warnings.');
```
Comment (Aenimus) [inline cli/src/commands/feature-flag/commands/create.ts:32]: same as above

File: cli/src/commands/feature-flag/commands/delete.ts [hunk 7]
```diff
@@ -14,6 +14,7 @@ export default (opts: BaseCommandOptions) => {
   command.argument('<name>', 'The name of the feature flag to delete.');
   command.option('-n, --namespace [string]', 'The namespace of the feature flag.');
   command.option('-f --force', 'Flag to force the deletion (skip confirmation).');
+  command.option('--suppress-warnings', 'This flag suppresses the composition warnings.');
```
Comment (Aenimus) [inline cli/src/commands/feature-flag/commands/delete.ts:17]: All the places we describe this flag. :D

File: controlplane/src/core/composition/composer.ts [hunk 8]
```diff
@@ -524,20 +529,31 @@ export class Composer {
...
-        const { federationResult: result, errors, federationResultContainerByContractName } = federationResultContainer;
+        const {
+          federationResult: result,
+          errors,
+          federationResultContainerByContractName,
+          warnings,
+        } = federationResultContainer;
 
-        composedGraphs.push(mapResultToComposedGraph(graph, subgraphs, errors, result));
+        composedGraphs.push(mapResultToComposedGraph(graph, subgraphs, errors, result, warnings));
```
Comment (Aenimus) [inline controlplane/src/core/composition/composer.ts:539]: `warnings` might need `|| []` for older versions of wgc?

Comment (Aenimus) [reply controlplane/src/core/composition/composer.ts:539]: NVM I see you handle it in `mapResultToComposedGraph` 👍

File: studio/src/components/federatedgraphs-cards.tsx [hunk 9]
```diff
@@ -56,7 +56,7 @@ import {
 } from "./ui/tooltip";
 import { useToast } from "./ui/use-toast";
 import { useMutation } from "@connectrpc/connect-query";
-import { GraphContext } from '@/components/layout/graph-layout';
+import { GraphContext } from "@/components/layout/graph-layout";
```
Comment (Aenimus) [inline studio/src/components/federatedgraphs-cards.tsx:59]: Weird. I thought prettier wanted single quotations?

File: studio/src/components/composition-errors-banner.tsx [hunk 10]
```diff
@@ -13,13 +13,13 @@ export const CompositionErrorsBanner = ({
...
       className={cn(
-        "mb-3 flex items-center justify-between space-x-2.5 rounded-lg border border-red-600 p-2 px-4 text-red-600 dark:border-red-900",
+        "mb-3 flex items-center justify-between space-x-2.5 rounded-lg border border-destructive p-2 px-4 text-destructive dark:border-red-900",
         className,
       )}
     >
       <div className="flex items-center justify-between space-x-2.5">
         <div>
-          <BoltSlashIcon className="h-5 w-5 text-red-500" />
+          <BoltSlashIcon className="h-5 w-5 text-destructive" />
```
Comment (Aenimus) [inline studio/src/components/composition-errors-banner.tsx:22]: Is it correct that both `text/border-red-500` and `text/border-red-600` use `text-destructive`?

File: proto/wg/cosmo/platform/v1/platform.proto [hunk 11]
```diff
@@ -2137,6 +2158,7 @@ message CreateFeatureFlagResponse {
   Response response = 1;
   repeated CompositionError composition_errors = 2;
   repeated DeploymentError deployment_errors = 3;
+  repeated CompositionWarning compositionWarnings = 4;
```
Comment (Aenimus) [inline proto/wg/cosmo/platform/v1/platform.proto:2161]: Please use snake_case for consistency :kekw:

Comment (Aenimus) [reply proto/wg/cosmo/platform/v1/platform.proto:2161]: Oh god they're all mixed up. OK NVM.

**Review by Aenimus — APPROVED** [review]
LGTM

File: cli/src/commands/contract/commands/update.ts (hunk 4 above)
Comment (JivusAyrus) [reply cli/src/commands/contract/commands/update.ts:106]: ?

File: cli/src/commands/contract/commands/update.ts (hunk 5 above)
Comment (JivusAyrus) [reply cli/src/commands/contract/commands/update.ts:94]: No, suppress warnings is a bool and composition warnings are from controlplane

File: cli/src/commands/contract/commands/create.ts (hunk 1 above)
Comment (JivusAyrus) [reply cli/src/commands/contract/commands/create.ts:153]: Done

File: cli/src/commands/contract/commands/create.ts (hunk 2 above)
Comment (JivusAyrus) [reply cli/src/commands/contract/commands/create.ts]: Done

File: cli/src/commands/contract/commands/update.ts (hunk 3 above)
Comment (JivusAyrus) [reply cli/src/commands/contract/commands/update.ts]: Done

File: cli/src/commands/contract/commands/update.ts (hunk 4 above)
Comment (JivusAyrus) [reply cli/src/commands/contract/commands/update.ts]: Done

File: cli/src/commands/feature-flag/commands/create.ts (hunk 6 above)
Comment (JivusAyrus) [reply cli/src/commands/feature-flag/commands/create.ts]: Done

File: cli/src/commands/feature-flag/commands/delete.ts (hunk 7 above)
Comment (JivusAyrus) [reply cli/src/commands/feature-flag/commands/delete.ts]: Done

File: cli/src/commands/contract/commands/update.ts (hunk 5 above)
Comment (Aenimus) [reply cli/src/commands/contract/commands/update.ts:107]: I'm an idiot.
